src/librustc_mir/interpret/eval_context.rs

```diff
@@ -252,17 +252,22 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
         &self,
         substs: T,
     ) -> EvalResult<'tcx, T> {
+        if !substs.needs_subst() {
+            return Ok(self.tcx.erase_regions(&substs));
+        }
+        // Stack frames beyond the first stackframe have their own set of substs and it would be
+        // absolutely wrong to use them for any further substitutions. So we just yield the substs
+        // after region erasing and let the caller figure out if that is good enough for them.
+        if self.stack.len() > 1 {
+            return Ok(self.tcx.erase_regions(&substs));
+        }
         match self.stack.last() {
             Some(frame) => Ok(self.tcx.subst_and_normalize_erasing_regions(
                 frame.instance.substs,
                 self.param_env,
                 &substs,
             )),
```

Comment:
New rounds of reviews: I'm pretty sure everything should use *at least* one frame, or you should keep a "top-level `substs" just like `param_env`.

Reply:
I went with the top level substs, because adding a new frame is way too expensive (not just in perf but also in "useless" API complexity) in some cases. We may be able to clean that up in the future.